packages/worker/src/workflow/reusable-vm.ts

```diff
@@ -0,0 +1,250 @@
+import assert from 'assert';
+import { AsyncLocalStorage } from 'async_hooks';
+import vm from 'vm';
+import { gte } from 'semver';
+import * as internals from '@temporalio/workflow/lib/worker-interface';
+import { IllegalStateError } from '@temporalio/common';
+import { Workflow, WorkflowCreateOptions, WorkflowCreator } from './interface';
+import { WorkflowBundleWithSourceMapAndFilename } from './workflow-worker-thread/input';
+import { BaseVMWorkflow, globalHandlers, setUnhandledRejectionHandler } from './vm-shared';
+
+// Thanks MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze
+function deepFreeze(object: any) {
+  // Retrieve the property names defined on object
+  const propNames = Object.getOwnPropertyNames(object);
+
+  // Freeze properties before freezing self
+  for (const name of propNames) {
+    const value = object[name];
+
+    if (value && typeof value === 'object') {
+      try {
+        deepFreeze(value);
+      } catch (err) {
+        // This is okay, there are some typed arrays that cannot be frozen (encodingKeys)
+      }
+    } else if (typeof value === 'function') {
+      Object.freeze(value);
+    }
+  }
+
+  return Object.freeze(object);
+}
+
+/**
+ * A WorkflowCreator that creates VMWorkflows in the current isolate
+ */
+export class ReusableVMWorkflowCreator implements WorkflowCreator {
+  /**
+   * TODO(bergundy): Get rid of this static state somehow
+   */
+  private static unhandledRejectionHandlerHasBeenSet = false;
+  static workflowByRunId = new Map<string, ReusableVMWorkflow>();
+
+  readonly hasSeparateMicrotaskQueue: boolean;
+
+  /**
+   * Optional context - this attribute is deleted upon on {@link destroy}
+   *
+   * Use the {@link context} getter instead
+   */
+  _context?: vm.Context;
+  /**
+   * Store the global object keys we want to share between contexts
+   */
+  readonly contextKeysToPreserve: Set<string>;
+
+  constructor(
+    script: vm.Script,
+    public readonly workflowBundle: WorkflowBundleWithSourceMapAndFilename,
+    public readonly isolateExecutionTimeoutMs: number
+  ) {
+    if (!ReusableVMWorkflowCreator.unhandledRejectionHandlerHasBeenSet) {
+      setUnhandledRejectionHandler((runId) => ReusableVMWorkflowCreator.workflowByRunId.get(runId));
+      ReusableVMWorkflowCreator.unhandledRejectionHandlerHasBeenSet = true;
+    }
+
+    // https://nodejs.org/api/vm.html#vmcreatecontextcontextobject-options
+    // microtaskMode=afterEvaluate was added in 14.6.0
+    this.hasSeparateMicrotaskQueue = gte(process.versions.node, '14.6.0');
```

Comment:
Can remove, <14.18 is no longer supported

Reply:
We should probably error on unsupported node versions.